tools/myexpect.1
.\" Automatically generated by Pod::Man version 1.02
.\" Mon Sep 24 19:22:33 2001
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` `
.    ds C' '
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
.    .
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "MYEXPECT 1"
.TH MYEXPECT 1 "perl v5.6.0" "2001-09-24" "User Contributed Perl Documentation"
.UC
.SH "NAME"
myexpext.pl: verifies output from program tests
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\f(CW\*(C`expect(FILE,MESSAGE,patlist)\*(C'\fR verifies that each pattern in
\&\f(CW\*(C`patlist\*(C'\fR matches lines in \s-1FILE\s0.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The simplest way of using \f(CW\*(C`expect()\*(C'\fR is to write a list of patterns
that should be found in the output of the test. Suppose the output
contains:
.PP
.Vb 8
\&  Output run of program QBFG running on day Sun Apr  8 08:48:23 ART 2001
\&  ..... more lines here
\&  Total volume: 34.56
\&  more lines here...
\&  Total area: 23.43
\&  more lines here...
\&  Total impedance: 46.4
\&  more lines here...
.Ve
You yount to chech that the figures in lines \f(CW\*(C`Total...\*(C'\fR are precise
to the first digit. The ouput is in file \f(CW\*(C`QBFG.out\*(C'\fR and you write a
small perl program like this
.PP
.Vb 1
\&   #!/usr/bin/perl
.Ve
.Vb 1
\&   require 'myexpect.pl';
.Ve
.Vb 5
\&   expect("QBFG.out","Check on ouput of QBFG.out",<<'EOT');
\&   Total volume: 34.5
\&   Total area: 23.4
\&   Total impedance: 4
\&   EOT
.Ve
.Vb 1
\&   final_check();
.Ve
In the default mode, \f(CW\*(C`expect()\*(C'\fR takes the first pattern at a time
and starts scanning the file from the beginning, each lie at a time
until it finds a line that matches the pattern. Patterns are the usual
Perl patterns. So that remember to escape asterisks '\f(CW\*(C`*\*(C'\fR', question
marks '\f(CW\*(C`?\*(C'\fR', dots and others. You can leave the dot unescaped since
it matches itself, but the pattern is less strict (dot matches any
other character also).  Normally, when entering patterns with a \fIhere
in\fR document, as in the previous example, you protect the backslash
characters in the pattern list using quotes in the \f(CW\*(C`'EOT'\*(C'\fR
terminator.
.PP
If the pattern is not found an error is reported and the test is
counted as a failure. If a line matching is found, \f(CW\*(C`expect()\*(C'\fR takes
the folowing pattern and continue scanning the file from the line
following the previous match. If all the patterns are matched, then
the test is counted as a succes. If \f(CW\*(C`FILE\*(C'\fR can't be opened, then this
is reported separately from error. The final count is reported by a
call to \f(CW\*(C`final_check()\*(C'\fR.
.PP
You can alter this default behavior adding \fImagic tokens\fR in
the pattern list. The magic tokens are
.Ip "_\|_REWIND_\|_" 4
.IX Item "__REWIND__"
Rewind the file, i.e. scan for the next match starting from the
beginning of the file, rather than from the last match. This is useful
when you don't know exactly the order in which the lines will appear.
For instance file
.Sp
.Vb 5
\&   #------ contents of file test1.out
\&   line at the beginning
\&   ...
\&   other line 
\&   ...
.Ve
matches the following call
.Sp
.Vb 5
\&   expect("test1.out","Check on ouput of test1.out",<<'EOT');
\&   other line 
\&   __REWIND__
\&   line at the beginning
\&   EOT
.Ve
thanks to the presence of the \f(CW\*(C`_\|_REWIND_\|_\*(C'\fR directive. 
.Ip "_\|_BACKWARD_\|_" 4
.IX Item "__BACKWARD__"
Scan the file backward for the next and subsequent patterns.
.Ip "_\|_FORWARD_\|_" 4
.IX Item "__FORWARD__"
Cancel the \f(CW\*(C`_\|_BACKWARD_\|_\*(C'\fR directive and continue scanning forward. 
.Ip "_\|_NO_SKIP_\|_" 4
.IX Item "__NO_SKIP__"
In no-skip mode the following pattern is found to match with exactly
the following line, rather to scan the file from the following line
down. 
.Ip "_\|_SKIP_\|_" 4
.IX Item "__SKIP__"
Return to skip mode. 
.Ip "_\|_EXACT_MATCH_\|_" 4
.IX Item "__EXACT_MATCH__"
Pattern has to match exactly the beginning of the line. (No special 
meaning for characters like \f(CW\*(C`.\*(C'\fR, \f(CW\*(C`*\*(C'\fR, etc...
.Ip "_\|_REGEXP_MATCH_\|_" 4
.IX Item "__REGEXP_MATCH__"
Go back to regexp match. 
.Sh "Sections"
.IX Subsection "Sections"
Sometimes it is useful to divide tests into sections. For this call
\&\f(CW\*(C`begin_section("section name")\*(C'\fR before each section, and
\&\f(CW\*(C`end_section()\*(C'\fR at the end.  All enclosed calls to \f(CW\*(C`expect()\*(C'\fR are
assumed to be in the same logical section of tests and a summary is
reported for that section.
.SH "AUTHOR"
.IX Header "AUTHOR"
Mario A. Storti <mstorti@intec.unl.edu.ar>


ChangeLog
2001-08-07  mstorti@spider  <mstorti@spider>

	* test/lupart/Makefile, test/lupart/check_part.m,
	test/lupart/wallke.epl, test/runtests.pl, src/readmesh.cpp: Adding
	tests for IISD bug

	* test/lupart/Makefile, test/lupart/wallke.epl, src/distmat.cpp,
	src/lusubd.cpp, applications/ns/ns.cpp: Fixed bug for IISD

	* test/lupart/Makefile, src/buffpack.h, src/readmesh.cpp,
	doc/Makefile: Backup prior to incorporate a fix to a bug in IISD.

2001-08-06  mstorti@spider  <mstorti@spider>

	* src/buffpack.h, src/distmat.cpp, src/readmesh.cpp:
	Make workaround for `namespace' not enabled on old gcc
	version in geronimo. (RH 5.2)

	* doc/OBJ/sched2.fig, doc/myexmapl.tex, doc/petscfem.in.tex, doc/pfmat.tex:
	Fixed some bugs in doc

	* tools/coall, test/lupart/Makefile, test/lupart/wallke.epl,
	test/runtests.pl, src/Makefile, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/elemset.cpp, src/fastmat2.h, src/lusubd.cpp,
	src/pfmat.h, src/readmesh.cpp, applications/ns/wallke.cpp,
	applications/advdif/advective.h, Makefile.base, TODO,
	doc/manual/Makefile, doc/manual/doc.dxx, doc/OBJ/iisd.fig,
	doc/OBJ/iisd2.fig, doc/OBJ/iisd3.fig, doc/OBJ/sched.fig,
	doc/OBJ/sched2.fig, doc/Makefile, doc/idmap.tex, doc/layout.tex,
	doc/odocadvdif.tex, doc/odocelems.tex, doc/odociisd.tex,
	doc/odocns.tex, doc/odocrmsh.tex, doc/petscfem.in.tex,
	doc/pfmat.tex: Added doc for PFMat, IISDMat, DistMap, DistMat and
	DistMatrix.

2001-08-05  mstorti@spider  <mstorti@spider>

	* doc/pfmat.tex:
	Doc for the PFMat abtract interface and in particular de IISD
	implementation.

2001-08-04  mstorti@spider  <mstorti@spider>

	* test/lupart/check_peri.m: Checks the periodic test

	* src/distmat.cpp: *** empty log message ***

2001-08-03  mstorti@spider  <mstorti@spider>

	* VERSION, save.log: Administrative files

	* src/readmesh.cpp.particion:
	Hitch-hiking partitioning was incorporated to `read_mesh'

	* test/lupart/Makefile, test/lupart/mkmesh.m,
	test/lupart/wallke.epl, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/lusubd.cpp: Fixed a bug in
	DistMap<int,Row>::unpack().  `row' has to be cleared before
	entering values.

	* TODO: order headers

2001-08-02  mstorti@spider  <mstorti@spider>

	* test/distmap/Makefile, test/distmap/distmat.cpp, src/distmap.h,
	src/distmat.cpp, src/distmat.h, src/lusubd.cpp, src/pfmat.h:
	Including the distributed matrix to IISD

	* test/runtests.pl, src/Makefile, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/lusubd.cpp, src/pfmat.h, src/utils.cpp,
	src/utils.h: Adding distributed matrix class DistMatrix to IISDMat

	* test/distmap/distmat.cpp: Moved code to src/distmat.cpp

	* test/distmap/Makefile: Adding DistMatrix to the IISDMat class.

2001-08-01  mstorti@spider  <mstorti@spider>

	* src/distmat.h: The distributed matrix class

	* src/distmat.cpp: The Distributed Matrix class

	* src/buffpack.h: The BufferPack class allows to pack standard and
	builtins types in a buffer.

	* src/maximizr.h: The maximizer class is a template that remembers
	the maximum of a set of values.

	* src/elemset.cpp, src/lusubd.cpp, src/readmesh.cpp, README,
	README.html: Working on `distmat'.  Backup after incorporating
	changes from some work in geronimo.

	* test/lupart/Makefile, test/distmap/distmat.cpp, src/Makefile, src/readmesh.cpp:
	Working on `distmat'.
	Backup before proceeding to install `distmat' on the IISDMat/PFMat'rices.

	* test/lupart/wallke.epl:
	removed old banner referring to `octave_export_vars'

	* test/distmap/Makefile, test/distmap/distmat.cpp, test/runtests.pl:
	Working on distributed matrix DistMap< int, map<int,double> >

2001-07-31  mstorti@spider  <mstorti@spider>

	* test/distmap/distmat.cpp:
	Distrubuted map<int,map<int,double>> i.e. a sparse matrix

	* test/distmap/distmap.cpp: Fixed a bug

	* tools/myexpect.pl:
	Modified so that without the -o option now sections are not print
	unless at least one test was found.

	* test/distmap/distmap.cpp: Distributed map is working.

	* src/distmap.h:
	Converted `char **buff' for unpacking to `const char *&buff'

	* src/distmap.h:
	Members that do not belong to this process are deleted now.

	* test/Makefile:
	Definition of variables has to be placed before the `includes'
	otherwise the conditional processing ('ifeq' and friend) don't work.

	* src/distmap.h: Distributed map is working basically.

	* src/Makefile: *** empty log message ***

	* test/distmap/Makefile, test/distmap/distmap.cpp, src/distmap.h:
	Debugging the distributed map class.
	Coded the send-receive paired communication.

	* src/distmap.h: Works, paired send-receive.

2001-07-30  mstorti@spider  <mstorti@spider>

	* test/distmap/Makefile, test/distmap/makefile.d, test/Makefile,
	src/distmap.h: Working on distributed map

	* src/Makefile, src/distmap.h, notes.txt, save.log:
	Working on distributed map class.

	* test/distmap/distmap.cpp, test/distmap/makefile.d:
	Workinf on distributed map class. Compiles.

2001-07-28  mstorti@spider  <mstorti@spider>

	* VERSION, save.log: Administrative files

	* test/wkepipe/Makefile, test/wallkechan/Makefile,
	test/wallke/Makefile, test/nonlr/Makefile, test/newff/Makefile,
	test/newff/data.m.epl, test/newff/newff.epl, test/lupart/Makefile,
	test/les/stabi3d/Makefile, test/les/stabi/Makefile,
	test/genload/Makefile, test/bug100/Makefile, test/runtests.pl,
	src/lusubd.cpp, src/pfmat.h, Makefile.base, notes.txt: Found a bug
	in IISD.  Added tests for advdif and ns with IISD.

	* src/lusubd.cpp:
	Fixed bug in lusubd.cpp. Will code it more efficiently.

2001-07-26  mstorti@spider  <mstorti@spider>

	* test/lupart/Makefile, applications/advdif/advdif.cpp, PROJECTS,
	TODO, notes.txt: Porting advdif to PFMat / IISD

2001-07-25  mstorti@spider  <mstorti@spider>

	* src/lusubd.cpp: In geronimo (RH 5.2.) doesn't support
	declarations of the form vector<int> flag0(dofmap->neq,0);

	* test/lupart/Makefile, test/lupart/wallke.epl, applications/ns/nsitetlesfm2.cpp, notes.txt:
	IISD works in multi-processor in RH 7.1 at least.


tools/myexpect.pl
#__INSERT_LICENSE__
#$Id: myexpect.pl,v 1.21 2007/02/20 14:57:47 mstorti Exp $

use English;
## position in count record
$OK= 1;
$NOT_OK= 2;
$CANT_OPEN= 3;

$COMPLAIN_ON_CANT_OPEN= 1 unless defined($COMPLAIN_ON_CANT_OPEN);

@stack=();
@output=();
@stack_output=();

$PRE = "{{";
$POST = "}}";
$SEP = "}{";
$COMMENT = "#>>";
$WD = ".";

## Allows determining machine architecture
$archi = `uname -i`;
chomp $archi;
$archi = 'i386' if $archi eq 'unknown';

sub P { print @_; }

sub read_file {
    my $file = shift();
    die "can't open file \"$file\"\n" unless  open FILE,$file;
    my $file = join "",(<FILE>);
    return $file;
}

sub printo_push {
    push @stack_output,[@output];
    @output = ();
}

sub printo_pop {
    my $last = pop @stack_output;
    flush() unless defined $last;
    unshift @output,@$last;
}

sub printo_discard {
    my $last = pop @stack_output;
    flush() unless defined $last;
    @output = ();
    unshift @output,@$last;
}

sub printo {
    push @output,@_;
}

sub flush {
    print @output;
    @output=();
}

sub match_exactly {
    my ($line,$patt) = @_;
    $np = length($patt);
    if (length($line) < $np) {return 0;};
    return (substr($line,0,$np) eq $patt);
}

sub match_regexp {
    my ($line,$patt) = @_;
    ## Parse pattern
    my $orig_patt = $patt;
    my $new_patt = "";
    my @conds = ();
    while ($patt =~ /$PRE/) {
	$patt = $POSTMATCH;
	$new_patt .= $PREMATCH;
	die "delimiters doesn't match in pattern: \"$orig_patt\"\n" 
	    unless $patt =~/$POST/;
	$cond = $PREMATCH;
	$patt = $POSTMATCH;
	my $subpat = ".*";
	if ($cond =~ /$SEP/) {
	    # my $condo = $cond;
	    $cond = $POSTMATCH;
	    $subpat = $PREMATCH;
	    # print "condo: $condo, subpat: $subpat, cond: $cond\n";
	}
	$new_patt .= "($subpat)";
	push @conds,$cond;
    }
    $new_patt .= $patt;
    # print "new_patt: $new_patt\n";
    # print "conds: <",join("><",@conds),">\n";
    my @matches = ($line =~ /$new_patt/);
    ## return 0 unless the line matched
    return 0 unless @matches;
    ## return 0 if there are conditions and they didn't match
    ## in number with the matches
    return 0 if @conds && @matches != @conds;
    ## check each condition if any
    for (my $j=0; $j<@conds; $j++) {
	$W = $matches[$j];
        ## print "W $W, cond ($conds[$j]) -> ",eval($conds[$j]),"\n";
	return 0 unless eval $conds[$j];
    }
    return 1;
}

sub read_output {
    my ($sal,$file) = @_;
    open (SAL,$file) || do {printo "can't open $file\n" 
				unless ! $COMPLAIN_ON_CANT_OPEN;
			    cant_open(); return 1; };
    @$sal=(<SAL>);
    close SAL;
    return 0;
}

sub expect {
    ($file,$descr,$pattern_list) = @_;
    @sal = ();
    return if $file && $file!~ m|/$| && read_output(\@sal,$file);
    $WD = $1 if $file =~ m|^(.*)/[^/]*$|;
    printo_push();
    printo "Testing: \"$descr\" on file \"$file\"...";
    printo "\n" if $opt_d;
    @pattern = split("\n",$pattern_list);
    $record=0;
    my $inc=1;
    my $skip=1;
    my $round_look=0;
    my $match_fun = \&match_regexp;
    my $file_changed = 0;
    while ($pattern=shift @pattern) {
	if ($pattern =~ /^$COMMENT/) { next; }
	if ($pattern =~ /^__ROUND_LOOK__$/) {
	    $round_look=1;
	    next;
	}
	if ($pattern =~ /^__NO_ROUND_LOOK__$/) {
	    $round_look=0;
	    next;
	}
	if ($pattern =~ /^__REWIND__$/) {
	    $inc=1;
	    $record=0;
	    next;
	}
	if ($pattern =~ /^__BACKWARD__$/) {
	    $inc=-1;
	    next;
	}
	if ($pattern =~ /^__FORWARD__$/) {
	    $inc=+1;
	    next;
	}
	if ($pattern =~ /^__SKIP__$/) {
	    $skip=1;
	    next;
	}
	if ($pattern =~ /^__NO_SKIP__$/) {
	    $skip=0;
	    next;
	}
	if ($pattern =~ /^__CONFIG__/) {
	    $rest = $POSTMATCH;
	    while ($rest =~ /^\s*(\S*)\s*(\S)/) {
		my $var = $1;
		my $delim = $2;
		$rest = $POSTMATCH;
		die "not valid var in config line var \"$var\", line: \"$pattern\"\n"
		    unless $var =~ /^(PRE|SEP|POST|WD|COMMENT)$/;
		die "unbalanced delimiters at config line: \"$pattern\"\n"
		    unless $rest =~ /^(.*?)$delim/;
		my $value = $1;
		$rest = $POSTMATCH;
		${$var} = $value;
		## print "setting \$${var} = $value\n";
	    }
	    die "can't parse config line: \"$pattern\"\n" unless $rest =~ /^\s*$/;
	    next;
	}
	if ($pattern =~ /^__SWITCH_FILE__\s*(\S*)\s*$/) {
	    $file = $1;
	    $file = "$WD/$file" unless $file=~ m|^/|;
	    @sal = ();
	    read_output(\@sal,$file);
	    $record=0;
	    $file_changed = 1;
	    next;
	}
	if ($pattern =~ /^__EXACT_MATCH__$/) {
	    $match_fun = \&match_exactly;
	    next;
	}
	if ($pattern =~ /^__REGEXP_MATCH__$/) {
	    $match_fun = \&match_regexp;
	    next;
	}
	$record=0 if $round_look;
	printo "trying pattern: \"$pattern\"...  \n" if $opt_d;
	while ($record<=$#sal) {
	    $_ = $sal[$record];
	    $record += $inc;
	    printo "$record: $_" if $opt_d;
	    do {chomp; 
		printo "        -> found: \"$_\"\n" if $opt_d; 
		goto NEXT;} if &{$match_fun}($_,$pattern);
	    last unless $skip;
	}
	not_ok();
	printo "not OK.\n        --->  Couldn't find \"$pattern\"\n";
	printo "    [on file $file]\n" if $file_changed;
	return;
      NEXT:;
    }
    printo "OK. \n";
    ok();
}

sub inc {
    my $f = shift();
    if ($#stack<0) {
	push @stack,["--- Tests ---",0,0,0];
    }
    my $t = pop @stack;
    $t->[$f]++;
    push @stack,$t;
}

sub ok { 
    if ($opt_n) { 
	printo_discard(); 
    } else {
	printo_pop(); 
    }
    inc($OK);
}
sub not_ok { printo_pop(); inc($NOT_OK);}
sub cant_open { printo_pop(); inc($CANT_OPEN);}

sub begin_section {
    my $s=shift();
    push @stack,[$s,0,0,0];
}

sub end_section {
    my $t = pop @stack;
    my $total = $t->[1]+$t->[2]+$t->[3];
    my $total_open = $t->[$OK] + $t->[$NOT_OK];
    my $print_sec = !$opt_n && ($total_open || $COMPLAIN_ON_CANT_OPEN)
	|| $opt_n && $t->[$NOT_OK];
    if ($print_sec) {
	print "--------\nStart section: \"$t->[0]\"\n";
	flush();
	print "Summary: \"$t->[0]\"",
	" -- OK: $t->[$OK].  Not OK: $t->[$NOT_OK]. ",
	"Couldn't open: $t->[$CANT_OPEN]. Total: $total\n";
    }
    if ($#stack>=0) {
	my $tt = pop @stack;
	$tt->[$OK] += $t->[$OK];
	$tt->[$NOT_OK] += $t->[$NOT_OK];
	$tt->[$CANT_OPEN] += $t->[$CANT_OPEN];
	push @stack,$tt;
    }
}

sub final_check {
    while ($#stack>=0) {
	end_section();
    }
}

=head1 NAME

myexpext.pl: verifies output from program tests

=head1 SYNOPSIS

C<expect(FILE,MESSAGE,patlist)> verifies that each pattern in
C<patlist> matches lines in FILE.

=head1 DESCRIPTION

The simplest way of using C<expect()> is to write a list of patterns
that should be found in the output of the test. Suppose the output
contains:

  Output run of program QBFG running on day Sun Apr  8 08:48:23 ART 2001
  ..... more lines here
  Total volume: 34.56
  more lines here...
  Total area: 23.43
  more lines here...
  Total impedance: 46.4
  more lines here...

You want to check that figures in lines C<Total...> are precise
to the first digit. The output is in file C<QBFG.out> and you write a
small perl program like this

   #!/usr/bin/perl
   
   require 'myexpect.pl';
   
   expect("QBFG.out","Check on output of QBFG.out",<<'EOT');
   Total volume: 34.5
   Total area: 23.4
   Total impedance: 4
   EOT

   final_check();

In the default mode, C<expect()> takes the first pattern at a time
and starts scanning the file from the beginning, each line at a time
until it finds a line that matches the pattern. Patterns are the usual
Perl patterns. So that remember to escape asterisks 'C<*>', question
marks 'C<?>', dots and others. You can leave the dot unescaped since
it matches itself, but the pattern is less strict (dot matches any
other character also).  Normally, when entering patterns with a I<here
in> document, as in the previous example, you protect the backslash
characters in the pattern list using quotes in the C<'EOT'>
terminator.

If the pattern is not found an error is reported and the test is
counted as a failure. If a line matching is found, C<expect()> takes
the following pattern and continue scanning the file from the line
following the previous match. If all the patterns are matched, then
the test is counted as a success. If C<FILE> can't be opened, then this
is reported separately from error. The final count is reported by a
call to C<final_check()>.

=head2 Special directives

You can alter this default behavior adding I<special directives> in
the pattern list. They are

=over 4

=item C<__REWIND__>

Rewind the file, i.e. scan for the next match starting from the
beginning of the file, rather than from the last match. This is useful
when you don't know exactly the order in which the lines will appear.
For instance file

   #------ contents of file test1.out
   line at the beginning
   ...
   other line 
   ...

matches the following call

   expect("test1.out","Check on output of test1.out",<<'EOT');
   other line 
   __REWIND__
   line at the beginning
   EOT
  
thanks to the presence of the L<__REWIND__> directive. 

=item C<__ROUND_LOOK__>

Rewind the file after each line, i.e. for each pattern look for it all
around the output. (Be careful for )

=item C<__BACKWARD__>

Scan the file backward for the next and subsequent patterns.

=item C<__FORWARD__>

Cancel the L<__BACKWARD__> directive and continue scanning forward. 

=item C<__NO_SKIP__>

In no-skip mode the following pattern line is checked to match with
exactly the following output line, rather to scan the whole output
file from the following line down.

=item C<__SKIP__>

Return to skip mode. 

=item C<__EXACT_MATCH__>

Pattern has to match exactly the beginning of the line. (No special 
meaning for characters like C<.>, C<*>, etc...

=item C<__REGEXP_MATCH__>

Go back to regexp match. 

=item C<__SWITCH_FILE__ <file>>

Stop reading this output file and switch over to C<file>. File 
name is taken absolute if starts with C</> otherwise relative to the 
current working directory. The working directory is taken from the
first file entered or changed with the C<<< __CONFIG_ WD "<dir>" >>> 
(directive. Example:
 
   __SWITCH_FILE__ foodir/foofile.out

=item C<__CONFIG__>

Allows changing configuration variables. Syntax is

      __CONFIG__  var "value" ...

Set configuration variable C<$var> to C<value>. Current configuration
variables are

=over 8

=item C<PRE>, C<SEP>, C<POST>

The patterns that matches for the start, end and separator 
of an embedded block. (See below)

=item C<COMMENT>

The pattern that, when at the start of a line means a
comment. Default: C<<< #>> >>>.

=item C<WD>

The working directory.

=back

The double quote delimiter around C<value> may be changed by any
non-blank character. Examples:

   __CONFIG__ PRE "<<" SEP "><" POST ">>" WD "foodir/bardir" COMMENT "#"
   __CONFIG__ PRE COMMENT |#:|

The last one sets comment start to C<#:>. 

=back

=head2 Embedded blocks

You can embed Perl code blocks in the pattern for checking the result
of the matches themselves. The syntax is C<{{CONDITION}}> or
C<{{PATTERN}{CONDITION}}>. In the first form the C<{{...}}> block is
replaced with C<(.*)>. If the output line matches, then every
C<CONDITION> is evaluated and the line matches if every condition
returns true. Inside the condition the special variable C<$W> takes 
the value of the matching string. Example: the following output lines

  Total mass: 0.356, 
    max density: 0.48956, min density: 0.001234

match the following pattern lines

  Total mass: {{abs($W-0.355)<2e-3}}, 
    max density: {{abs($W-0.4)<0.1}}, min density: {{$W<2e-3}}

In the second form the C<PATTERN> section allows to specify the pattern 
which replaces the block. Example: the output line

  Processed case: ABC890

matches

  Processed case: {{[A-Z]*}{$W eq 'ABC'}}{{\d*}{$W>800 && $W < 900}}

The syntax of the block may be changed with the L<C<__CONFIG__>>
directive. The corresponding variables are C<PRE>, C<SEP> y <POST>
(mnemonic: prefix, separator and postfix). Possible choices are

  #>> Following three examples use matching delimiters (like <>, () or[])
  #>> warning: angles (<>) may collide with comparison expressions
  __CONFIG__ PRE "((" SEP ")(" POST "))"
  __CONFIG__ PRE "<<" SEP "><" POST ">>"
  __CONFIG__ PRE "[[" SEP "][" POST "]]"
  #>> This is very simple
  __CONFIG__ PRE "{" SEP "," POST "}"
  #>> Another simple one
  __CONFIG__ PRE "<" SEP "|" POST "}"

In order to avoid collision you can increase the delimiter levels, e.g.

  #>> Very paranoid
  __CONFIG__ PRE "{{{{" SEP "}}{{" POST "}}}}"
  #>> Combines with colon
  __CONFIG__ PRE "<<:" SEP ":><:" POST ":>>"

=head2 Sections

Sometimes it is useful to divide tests into sections. Start sections
with C<begin_section(<section name>)>, and end with C<end_section()>.
All enclosed calls to C<expect()> are assumed to be in the same
logical section of tests and a summary is reported for that section.
Example:

 begin_section("Navier-Stokes tests");
 expect("NS/output1.txt","NS Test 1","NS Test 1 OK");
 expect("NS/output1.txt","NS Test 2","NS Test 2 OK");
 expect("NS/output1.txt","NS Test 3","NS Test 3 OK");
 end_section();

 begin_section("Electro-magnetic tests");
 expect("EM/output1.txt","EM Test 1","EM Test 1 OK");
 expect("EM/output1.txt","EM Test 2","EM Test 2 OK");
 expect("EM/output1.txt","EM Test 3","EM Test 3 OK");
 expect("EM/output1.txt","EM Test 4","EM Test 4 OK");
 end_section();

=head1 AUTHOR

Mario A. Storti <mario.storti@gmail.com>

=cut

1;    
